cpp/src/gandiva/gdv_string_function_stubs.cc

```diff
@@ -264,6 +265,304 @@ const char* gdv_fn_lower_utf8(int64_t context, const char* data, int32_t data_le
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 4);
+        tmp += 4;
+      } else if (char_len == 3) {
+        const auto* in_char = (const uint8_t*)(data + i);
+
+        // Decode the multibyte character
+        bool is_valid_utf8_char =
+            arrow::util::UTF8Decode((const uint8_t**)&in_char, &char_codepoint);
+
+        // If it is an invalid utf8 character, UTF8Decode evaluates to false
+        if (!is_valid_utf8_char) {
+          gdv_fn_set_error_for_invalid_utf8(context, data[i]);
+          *out_len = 0;
+          return "";
+        }
+
+        char hex_string[20];
+
+        snprintf(hex_string, z, "%.8X", char_codepoint);
+
+        for (int32_t xx = 0; xx < 8; xx += 4) {
+          char pp = hex_string[xx];
+          hex_string[xx] = hex_string[xx + 2];
+          hex_string[xx + 2] = pp;
+
+          pp = hex_string[xx + 1];
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 8);
+        tmp += 8;
+      } else if (char_len == 4) {
+        const auto* in_char = (const uint8_t*)(data + i);
+
+        // Decode the multibyte character
+        bool is_valid_utf8_char =
+            arrow::util::UTF8Decode((const uint8_t**)&in_char, &char_codepoint);
+
+        // If it is an invalid utf8 character, UTF8Decode evaluates to false
+        if (!is_valid_utf8_char) {
+          gdv_fn_set_error_for_invalid_utf8(context, data[i]);
+          *out_len = 0;
+          return "";
+        }
+
+        char hex_string[20];
+
+        snprintf(hex_string, z, "%.8X", char_codepoint);
+
+        for (int32_t xx = 0; xx < 8; xx += 4) {
+          char pp = hex_string[xx];
+          hex_string[xx] = hex_string[xx + 2];
+          hex_string[xx + 2] = pp;
+
+          pp = hex_string[xx + 1];
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 8);
+        tmp += 8;
+      }
+    }
+    *out_len = static_cast<int32_t>(tmp - out);
+    return out;
+  }
+  return "";
+}
+
+// converts utf16 to utf8 encoding
+GANDIVA_EXPORT
+const char* gdv_fn_decode(int64_t context, const char* data, int32_t data_len,
+                          const char* charset, int32_t charset_len, int32_t* out_len) {
+  if (data_len == 0 || data_len % 4 != 0) {
+    *out_len = 0;
+    return "";
+  }
+
+  char* out =
+      reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, 32 * data_len));
+  if (out == nullptr) {
+    gdv_fn_context_set_error_msg(context, "Could not allocate memory for output string");
+    *out_len = 0;
+    return "";
+  }
+
+  std::string charset_str(charset, charset_len);
+
+  if (boost::iequals(charset_str, "UTF-16BE")) {
+    char* tmp = out;
+
+    for (int32_t i = 0; i < data_len; i += 2) {
+      if (data[i] == '0' && data[i + 1] == '0') continue;
```

Comment:
Why???

Reply:
Because if both are zero we have to ignore as we have to convert it from utf16 to utf8.

Reply:
Why ignore?

Reply:
Because they won't be of any use while converting back to utf8. I am getting a value then I typecast it.

Reply:
Can you explain this more clearly? Where is this "00" sequence coming from? Why would zero bytes be ignored? Multiples of 256 are usually valid Unicode characters. If you ignore some input, you will probably get desynchronized wrt. the original utf8 data.

Reply:
Utf8 codepoints are converted to Utf16 codepoints which means 1-2 bytes codepoints will be converted to 2 bytes and 3-4 bytes codepoints will be converted to 4 bytes. So suppose if there is the utf8 1 byte character then in utf16 it would be converted to 2 bytes which will be only possible by adding 0's so in the reverse process we have to ignore those 0's. Hope it makes sense to you.

Reply:
Say I have the unicode codepoint 256. Once encoded in utf-16-be, it gives the string "0100". Your code seems to ignore the two latter digits, doesn't it?

Reply:
Ok yeah got it thanks have corrected that now :-).